### PR TITLE
OmegaConf.to_yaml 

### LINKED ISSUE #263: OmegaConf.pretty()
Provide an alternative OmegaConf.pretty() to cfg.pretty() to enable pretty printing of duck-typed configs.

We can deprecate it in favor of:

```
OmegaConf.to_yaml(cfg, resolve: bool = False, sort_keys: bool = False)
```

### PR DESCRIPTION
**Pull request by pereman2**
Closes #263 .
- [x] Documentation
- [x] Jupyter notebook change

Pretty() instance method is changed to OmegaConf.to_yaml() static method.

### CODE REVIEW

File: omegaconf/base.py [hunk 1]
```diff
@@ -205,10 +205,6 @@ class Container(Node):
 
     _metadata: ContainerMetadata
 
-    @abstractmethod
-    def pretty(self, resolve: bool = False, sort_keys: bool = False) -> str:
-        ...
```
Comment (omry) [inline omegaconf/base.py:210]: You have just pissed off every single use of OmegaConf :)

deprecate it, have it issue a warning and call OmegaConf.to_yaml().
See other examples in the codebase. (look for "warn").

File: omegaconf/omegaconf.py [hunk 2]
```diff
@@ -568,6 +569,24 @@ def update(cfg: Container, key: str, value: Any = None) -> None:
             idx = int(last)
             root[idx] = value
 
+    @staticmethod
+    def to_yaml(cfg: Container, resolve: bool = False, sort_keys: bool = False) -> str:
```
Comment (omry) [inline omegaconf/omegaconf.py:573]: 1. I think this is a good opportunity to change the default of resolve to True. (but be sure to mention in in the deprecation warning and the docs).

2. Force named parameters for everything after cfg, see [this](https://stackoverflow.com/questions/2965271/forced-naming-of-parameters-in-python). This will ensure that it's easier to extend this function in the future without worrying about people calling the flags positionally.

File: tests/examples/test_dataclass_example.py [hunk 3]
```diff
@@ -159,7 +159,7 @@ def test_nesting() -> None:
   name: manager
   height: TALL
 """
-    assert conf.pretty() == expected
+    assert OmegaConf.to_yaml(conf) == expected
```
Comment (omry) [inline tests/examples/test_dataclass_example.py:162]: after you deprecate, add another test validating the deprecation behavior (see other similar tests).

File: omegaconf/base.py (hunk 1 above)
Comment (pereman2) [reply omegaconf/base.py:210]: Oh shit true i'll undo that change asap

Comment (omry) [reply omegaconf/base.py:210]: No rush :), this is not landed and released yet.

File: tests/examples/test_dataclass_example.py (hunk 3 above)
Comment (pereman2) [reply tests/examples/test_dataclass_example.py:162]: Gotta leave, I'll add documentation and change and change the jupyter notebook.

**Comment by pereman2** [comment]
Gotta leave, I'll add documentation and change and change the jupyter notebook.

File: omegaconf/basecontainer.py [hunk 4]
```diff
@@ -214,6 +215,34 @@ def convert(val: Node) -> Any:
...
+        interpolations are preserved
+        :param sort_keys: If True, will print dict keys in sorted order. default False.
+        :return: A string containing the yaml representation.
+        """
+        from omegaconf import OmegaConf
+
+        warnings.warn(
+            """
+            pretty() is deprecated, use OmegaConf.to_yaml() and resolve
+            now defaults to True (Since 2.0.1)
```
Comment (omry) [inline omegaconf/basecontainer.py:231]:
```suggestion
            pretty() is deprecated and will be removed in a future version.
            Use OmegaConf.to_yaml(). Please note that the default value for resolve has changed to True.
```

File: omegaconf/basecontainer.py [hunk 5]
```diff
@@ -214,6 +215,34 @@ def convert(val: Node) -> Any:
...
         assert False
 
+    def pretty(self, resolve: bool = False, sort_keys: bool = False) -> str:
+        """
+        returns a yaml dump of this config object.
+        :param resolve: if True, will return a string with the interpolations resolved, otherwise
+        interpolations are preserved
+        :param sort_keys: If True, will print dict keys in sorted order. default False.
+        :return: A string containing the yaml representation.
+        """
```
Comment (omry) [inline omegaconf/basecontainer.py]:
```suggestion
```

File: omegaconf/base.py [hunk 6]
```diff
@@ -205,6 +205,10 @@ class Container(Node):
 
     _metadata: ContainerMetadata
 
+    @abstractmethod
+    def pretty(self, resolve: bool = False, sort_keys: bool = False) -> str:
+        ...
+
```
Comment (omry) [inline omegaconf/base.py]: why?

File: omegaconf/basecontainer.py [hunk 7]
```diff
@@ -214,6 +215,34 @@ def convert(val: Node) -> Any:
...
+        )
+
+        container = OmegaConf.to_container(self, resolve=resolve, enum_to_str=True)
+        return yaml.dump(  # type: ignore
+            container,
+            default_flow_style=False,
+            allow_unicode=True,
+            sort_keys=sort_keys,
+            Dumper=get_omega_conf_dumper(),
+        )
```
Comment (omry) [inline omegaconf/basecontainer.py]: delegate to OmegaConf.to_yaml()

File: tests/test_basic_ops_dict.py [hunk 8]
```diff
@@ -119,50 +119,50 @@ def test_subscript_set_with_dot_warning_suppressed(recwarn: Any, mocker: Any) ->
     assert len(recwarn) == 0
```
Comment (omry) [inline tests/test_basic_ops_dict.py:119]: all pretty() tests except the deprecation one should move to test_omegaconf()

File: tests/test_basic_ops_list.py [hunk 9]
```diff
@@ -30,47 +30,47 @@ def test_list_of_dicts() -> None:
     assert c[1].key2 == "value2"
 
 
-def test_pretty_list() -> None:
```
Comment (omry) [inline tests/test_basic_ops_list.py:33]: same, this should now be tested in test_omegaconf()

File: tests/test_nodes.py [hunk 10]
```diff
@@ -353,22 +354,36 @@ def test_legal_assignment_enum(
...
+        match=re.escape(
+            """
+            pretty() is deprecated, use OmegaConf.to_yaml() and resolve
+            now defaults to True (Since 2.0.1)
+            """
+        ),
+    ):
+        assert c.pretty() == "foo: bar\n"
+
+
```
Comment (omry) [inline tests/test_nodes.py:386]: not sure this is the vest place for it. why test_nodes?

Comment (pereman2) [reply tests/test_nodes.py:386]: Since the abstract method was in Container I thought it would be better but now looking at it it doesn't make that much sense. I'll move everything to test_omegaconf as you said since it's where to_yaml is.

File: tests/examples/test_dataclass_example.py (hunk 3 above)
Comment (pereman2) [reply tests/examples/test_dataclass_example.py:162]: Do I move test_nesting too ? I testes more than to_yaml.

File: docs/notebook/Tutorial.ipynb [hunk 11]
```diff
@@ -528,8 +528,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Interpolations are evaluated lazily on field access.<br/>\n",
-    "Note below that when printed the interpolations are not resolved.<br/>\n",
+    "Note below that when printed the interpolations are resolved.<br/>\n",
     "They get resolved once you access them."
```
Comment (omry) [inline docs/notebook/Tutorial.ipynb:532]: Interpolations are evaluated lazily on field access. OmegaConf.to_yaml() is eagerly resolving the interpolations by default. You can change it by passing resolve=False.

Example without resolve=False.
Example with resolve=False

Comment (omry) [reply docs/notebook/Tutorial.ipynb:532]: The other related doc update here is also similarly confusing.

File: docs/notebook/Tutorial.ipynb [hunk 12]
```diff
@@ -521,15 +521,14 @@
     "\n",
     "OmegaConf support variable interpolation, Interpolations are evaluated lazily on access.\n",
     "\n",
-    "### Config node interpolation"
+    "## Config node interpolation"
```
Comment (omry) [inline docs/notebook/Tutorial.ipynb:524]: Please provide a screen shot, I am not good at reading json data.

File: docs/source/usage.rst [hunk 13]
```diff
@@ -322,8 +339,8 @@ The following example sets `12345` as the the default value for the `DB_PASSWORD
 
 Custom interpolations
 ^^^^^^^^^^^^^^^^^^^^^
-You can add additional interpolation types using custom resolvers.
-This example creates a resolver that adds 10 the the given value.
+You can add additional interpolation types using custom resolvers, which take strings as inputs.
+This example creates a resolver that adds 10 to the given value (note the need to cast it to `int`).
```
Comment (omry) [inline docs/source/usage.rst:343]: Is this not already committed in master? why is this here again?

File: docs/notebook/Tutorial.ipynb (hunk 12 above)
Comment (pereman2) [reply docs/notebook/Tutorial.ipynb]: image

File: omegaconf/basecontainer.py [hunk 14]
```diff
@@ -216,24 +215,20 @@ def convert(val: Node) -> Any:
...
-            sort_keys=sort_keys,
-            Dumper=get_omega_conf_dumper(),
+        warnings.warn(
+            """
+            pretty() is deprecated and will be removed in a future version.
+            Use OmegaConf.to_yaml. Please note that the default value for
+            resolve has changed to True.
+            """,
+            category=UserWarning,
+            stacklevel=2,
```
Comment (omry) [inline omegaconf/basecontainer.py:227]: why stackLevel 2?

File: tests/examples/test_dataclass_example.py (hunk 3 above)
Comment (omry) [reply tests/examples/test_dataclass_example.py:162]: no.

File: tests/test_basic_ops_list.py [hunk 15]
```diff
@@ -30,225 +22,6 @@ def test_list_of_dicts() -> None:
...
-                    {"foo": "${defaults.0.optimizer}_${defaults.1.dataset}"},
-                ]
-            },
-            [{"optimizer": "adam"}, {"dataset": "imagenet"}, {"foo": "adam_imagenet"}],
-            "defaults",
-            id="str_interpolation",
-        ),
-    ],
-)
-def test_iterate_list(input_: Any, expected: Any, list_key: str) -> None:
```
Comment (omry) [inline tests/test_basic_ops_list.py:102]: You are moving unrelated tests.
always review your own diffs before asking for someone else to review them.

File: docs/notebook/Tutorial.ipynb [hunk 16]
```diff
@@ -726,7 +733,7 @@
    ],
    "source": [
     "conf = OmegaConf.load('../source/env_interpolation.yaml')\n",
-    "print(conf.pretty(resolve=True))"
```
Comment (omry) [inline docs/notebook/Tutorial.ipynb:729]: This example was making a point of showing the effect of resolve.
you removed resolve, but now it's not making any point because resolve is True by default.

File: tests/test_omegaconf.py [hunk 17]
```diff
@@ -381,3 +398,650 @@ def test_is_issubclass() -> None:
...
+        print(t)
+        c = OmegaConf.create([t, 1])
+        expected = "- '%s'\n- 1\n" % t
+        assert OmegaConf.to_yaml(c) == expected
+
+
+def test_to_yaml_string_int_list() -> None:
+    c = OmegaConf.create(["1", 1])
+    expected = "- '1'\n- 1\n"
+    assert OmegaConf.to_yaml(c) == expected
```
Comment (omry) [inline tests/test_omegaconf.py:443]: I didn't realize we now have a zillion tests for pretrty.
1. create a new dedicated test file for it : test_to_yaml.py
2. use parameterize to reduce the number of tests. for example the two tests here are almost identical and can be easily pameterized. (do this in two diffs so it's easier to review. first diff moves all the to_yaml tests to one file and the second parameterize.
3.  I see a print here. remove it and other prints you may have added in tests.

File: docs/notebook/Tutorial.ipynb (hunk 16 above)
Comment (pereman2) [reply docs/notebook/Tutorial.ipynb:729]: Changed in the last commit, now the next cell shows wihout resolving.

File: omegaconf/basecontainer.py (hunk 14 above)
Comment (pereman2) [reply omegaconf/basecontainer.py:227]: I didn't give that much thought to it, I saw that select and update_node had the same parameters and all of them are in basecontainer so I did the same.

Comment (omry) [reply omegaconf/basecontainer.py:227]: Don't pass stackLevel at all unless you do give it some thought.

File: tests/test_basic_ops_dict.py [hunk 18]
```diff
@@ -119,50 +118,249 @@ def test_subscript_set_with_dot_warning_suppressed(recwarn: Any, mocker: Any) ->
...
-    expected = """hello: world
-list:
-- 1
-- 2
-"""
-    assert expected == c.pretty()
-    assert OmegaConf.create(c.pretty()) == c
+def test_instantiate_config_fails() -> None:
+    with pytest.raises(TypeError):
+        BaseContainer()  # type: ignore
```
Comment (omry) [inline tests/test_basic_ops_dict.py:123]: This is not a real change, you are just moving it around in the file.
these kind of changes are making it really hard to review the diff.
Please fix your diff. at that point it might be easier to revert the changes to the logs and move just the correct functions to the new file.

File: tests/test_nodes.py [hunk 19]
```diff
@@ -353,20 +354,19 @@ def test_legal_assignment_enum(
...
+        expected_warning=UserWarning,
+        match=re.escape(
+            """
+            pretty() is deprecated and will be removed in a future version.
+            Use OmegaConf.to_yaml. Please note that the default value for
+            resolve has changed to True.
+            """,
+        ),
+    ):
+        assert c.pretty() == "foo: bar\n"
```
Comment (omry) [inline tests/test_nodes.py:369]: this test makes no sense in test_nodes.
move to test_to_yaml.

**Comment by omry** [comment]
How will users know about this important change?

**Comment by pereman2** [comment]
Update NEWS.md ?

**Comment by omry** [comment]
you never update NEWS directly, but close: create a news fragment.

**Comment by omry** [comment]
having second thoughts about changing the default for resolve to True:
if your config has interpolations that can't be resolved I think it will blow up.

**Comment by pereman2** [comment]
Ok, if you want I can turn it back. 
By interpolation that can't be resolved you mean a human error? if that's the case maybe it should raise and exception. Otherwise what type of interpolation cannot be resolved(if you have a snippet cool, if not i'll investigate it)? 
I'm going to bed now, I'll look into it tomorrow. 😃

**Comment by omry** [comment]
so remember that this is used primarily for debugging, you would not be happy to get an exception from a str() method due to an error in your data, right?

**Comment by pereman2** [comment]
True.

**Comment by omry** [comment]
Okay then, let's change the default of resolve back to False. remember to update the docs and news fragment.
